File: Sean的leetcode/Leetcode/Path Sum II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    void findPathSum(TreeNode *root, int sum, vector<vector<int> > &res, vector<int> &one_sol)
    {
        if (!root)
            return;
            
        if (root->val == sum && !root->left && !root->right)            
        {
            one_sol.push_back(root->val);
            res.push_back(one_sol);
            one_sol.pop_back();
            return;
        }
        
        one_sol.push_back(root->val);
        findPathSum(root->left, sum-root->val, res, one_sol);
        findPathSum(root->right, sum-root->val, res, one_sol);        
        one_sol.pop_back();                    
    }
    
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> >res;
        vector<int>one_sol;
        
        findPathSum(root, sum, res, one_sol);
        return res;
    }
};

File: Sean的leetcode/python/Path Sum II.py
###
# Time Complexity: O(nlogn) traverse n node, each node make copy of current path
# Space Complexity: O(nlogn) recursive stack
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, s):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        result = []
        self.dfs(root, s, [], result)
        return result

    def dfs(self, cur, s, one_sol, result):
        if not cur:
            return
        if not cur.left and not cur.right:
            if cur.val == s:
                result.append(one_sol + [cur.val])
            return

        self.dfs(cur.left, s - cur.val, one_sol + [cur.val], result)
        self.dfs(cur.right, s - cur.val, one_sol + [cur.val], result)